Step2 자판기 기본 동작

seizze opened the pull request:
### 고민한 점
지정 생성자에서 인스턴스 내부 프로퍼티를 다 정해버리는 것은 부적절한 것 같아서 매개변수를 통해 인자들을 전달받아 인스턴스를 생성하도록 하였습니다. 하지만 이렇게 되니 인스턴스를 생성할 때마다 타이핑을 너무 많이 하게 되는 것 같아서 구체적인 하위 클래스들에게는 매개변수 기본값을 제공하였습니다.

### 궁금한 점
상속 관계에서 `Equatable`을 어떻게 적합하게 구현해야 할지 잘 떠오르지 않아서, 테스트 시에 같은지를 비교할 때 인스턴스가 같은지를 확인하는 `===` 연산자를 사용했는데, 이 방법이 맞는지 아니면 모든 타입들에 `Equatable`을 구현한 뒤에 테스트시에 타입 캐스팅등의 작업을 하여 `==` 연산자를 이용하는 것이 맞는지 잘 모르겠습니다.

seizze committed:
feat: Implement milk that inherits beverage

seizze committed:
feat: Make flavored milks inherit milk

seizze committed:
feat: Add soft drink that inherits beverage

seizze committed:
feat: Make soft drinks inherit class soft drink

seizze committed:
feat: Add coffee that inherits beverage

seizze committed:
feat: Make caffeine drinks inherit class coffee

seizze committed:
feat: Remove some beverages

seizze committed:
feat: Add access control and interfaces to milk

seizze committed:
test: Test milk validation

seizze committed:
feat: Add access control and interfaces

Change some properties to private
Add some interfaces to coffee and soft drinks

seizze committed:
test: Test coffee interfaces

seizze committed:
feat: Provide default parameter values

seizze committed:
feat: Abstract beverages to inventory

seizze committed:
feat: Add closure invokers

seizze committed:
test: Test adding, filtering available items

seizze committed:
feat: Add inventory managing methods

seizze committed:
test: Test inventory methods

seizze committed:
fix: Fix block(.success) invocation when failed

seizze committed:
feat: Add vending machine managing methods

seizze committed:
test: Test vending machine methods

seizze committed:
test: Add Integration tests

seizze committed:
docs: Update README.md

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverages/ChocolateMilk.swift:
```diff
@@ -8,8 +8,20 @@
-    init() {
-        super.init(manufacturer: "매일유업", size: 300, price: 1500, name: "바나나는 원래 하얗다")
+final class ChocolateMilk: Milk {
+    private let containsMint: Bool
```
데이터를 표현하는 모델을 설계할 때 타입을 Bool 보다는 양을 숫자로 표현하는 게 확장하기 좋습니다.
Bool은 판단하는 기준에 따라 다른 결과가 나오도록 하기 어려우니까요. 참고하세요.

seizze replied:
넵 Bool보다 양을 숫자로 표현하는게 더 확장하기 좋다는 점 기억하겠습니다!
민트초코보다는 카카오 함유량이 더 적절한 프로퍼티인것 같아서 프로퍼티 이름과 타입 모두 바꿨습니다. 0ed05920

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverages/Coffee.swift:
```diff
@@ -0,0 +1,33 @@
+    }
+    
+    func isHot() -> Bool {
+        return temperature > 80
```
canReplaceEnergyDrinks() 이나 isHot() 에서 기준값을 하드코딩했는데 이것을 코드로 표현해보세요.
덧붙여서, 기준값도 매개변수로 넘길 수 있으면 어떨까요? 좀 더 유연해지지 않을까요?

seizze replied:
넵 기준값을 매개변수로 넘기도록 변경하였습니다. 35dbe265
테스트 코드도 이에 맞게 변경했는데 가독성 면에서 좀 더 나아진 것 같습니다. 65c02d13

godrm commented on VendingMachineApp/VendingMachineApp/Models/Beverages/Georgia.swift:
```diff
@@ -8,8 +8,28 @@
+        case americano
+    }
+    
+    private(set) var productType: ProductType
```
Flavor, ProductType 처럼 enum 타입을 사용할 때도 주의할 점이 있습니다.
enum 타입이 추가되면, 관련이 있어서 영향을 주는 곳이 많아지면 
enum 타입을 쓰기 보다는 enum 속성을 포함하는 하위 클래스를 만드는 게 낫습니다.
그러니까 OriginalGeorgia, LatteGeorgia, AmericanoGeorgia 클래스를 만드는 게 나을 수 있습니다. 
그래야 OriginalGeorgia만 가격을 바꾸거나 로직을 바꿀 때 코드가 OriginalGeorgia 클래스만 수정하면 되겠죠.
그렇지 않으면 로직를 처리하는 곳마다 비교문으로 .original 인지 판단해서 분기처리를 해줘야 합니다.

seizze replied:
넵 수정하였습니다. dc79da54
앞으로 enum을 사용할 때 영향력도 고려하여 사용하겠습니다!

godrm commented on VendingMachineApp/VendingMachineApp/Models/Inventory.swift:
```diff
@@ -0,0 +1,73 @@
+enum InventoryError: Error {
+    case insufficientBalance
+    case notInStock
+}
```
Inventory 관련 에러면 nested 하는 것과 어떤 차이가 있을까요?

seizze replied:
제품을 판매할 때 인벤토리에서 발생하는 에러들이므로 nested하는 것이 더 적절할 것 같아서 Inventory.SaleError로 접근하도록 수정하였습니다!! 4055cb98

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -9,13 +9,34 @@
     
     var stockList: String {
-        return "\(stock)"
+        return "\(inventory)"
```
이렇게 출력하는 것은 description을 활용하는 것은 좋은 접근이 아닙니다. 
항상 문자열로 출력한다는 보장이 없으니까요.
특히 앱에서 뷰를 만들어야 하는데, 문자열로 넘기는 것은 좋은 선택이 아닙니다. 바로 다음 단계만 가도 쓸모없어 질겁니다.
데이터 구조로 전달하는 것을 고려하세요.

seizze replied:
네 클로저를 통해 접근하도록 수정하였습니다! 39e9b4cc

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -9,13 +9,34 @@
 class VendingMachine {
-    private var stock = [Beverage]()
+    private var inventory = Inventory()
+    private(set) var balance = 0
```
balance나 price에 대한 Money 타입을 설계해보세요.
Int 보다 더 구체적으로 돈에 대한 로직을 담당하는 객체가 있는 게 좋을 것 같습니다.

seizze replied:
돈에 관한 로직을 담당하는 Cashier를 새로 만들고 그에 맞게 기능들을 분리하였습니다! 300a881d
Cashier 테스트 코드도 추가하였습니다. d8905edd

또한 Cashier로 돈에 관련한 로직들을 분리하다 보니, VendingMachine의 테스트가 어려워져서, Calculable 프로토콜을 선언한 뒤 이 프로토콜을 준수하는 Mock 객체를 만들어서 VendingMachine을 테스트해보려고 노력했습니다. f97cbab2, a7f81e61

godrm commented on VendingMachineApp/VendingMachineApp/Models/VendingMachine.swift:
```diff
@@ -9,13 +9,34 @@
+        inventory.add(beverage)
+    }
+    
+    func sell(beverage: String) {
```
음료를 구매할 때도 Beverage 타입을 그대로 활용할 수 있지 않나요? 문자열로 해야할 이유가 있나요?

seizze replied:
넵 수정하였습니다! 62d638e6

godrm reviewed (commented):
테스트 케이스가 있는 것은 매우 좋습니다. 
세부적인 부분에 대해 피드백을 확인해보세요.

godrm commented:
> 상속 관계에서 Equatable을 어떻게 적합하게 구현해야 할지 잘 떠오르지 않아서, 테스트 시에 같은지를 비교할 때 인스턴스가 같은지를 확인하는 === 연산자를 사용했는데, 이 방법이 맞는지 아니면 모든 타입들에 Equatable을 구현한 뒤에 테스트시에 타입 캐스팅등의 작업을 하여 == 연산자를 이용하는 것이 맞는지 잘 모르겠습니다.

상속 관계를 가진 클래스들 Equatable 은 간단하게 생각하면 각 클래스 기준에서 모든 속성이 같으면 되겠죠. 
인스턴스가 같은지 비교하는 것은 좋은 방법이 아닙니다. 그건 메모리 주소가 같은지 보는거니까 Equatable 아닙니다. 

Hashable 프로토콜에 대해서도 학습해보세요. 모든 속성을 비교하는 것보다 효과적인 방법을 제시해줍니다.

seizze committed:
refactor: Change to proper property

Change property containsMint(Bool) to cacaoContentRate(Int) which represents choco milk better

seizze committed:
refactor: Change some interfaces of beverages

Change threshold values to be passed from the outside

seizze committed:
test: Change to pass threshold value

seizze committed:
refactor: Change inventory error to nested type

seizze committed:
refactor: Abstract functions related to money

seizze committed:
test: Add cashier test methods

seizze committed:
refactor: Add protocol calculable

Make vending machine depends on calculable

seizze committed:
test: Add vending machine test methods

Test using mock cashier that conforms protocol calculable

seizze committed:
feat:  Make beverages conform protocol hashable

Make beverage and all its subclasses conform protocol hashable
Add functions to verify equality between instances of same types

seizze committed:
test: Improve tests using protocol equatable

seizze committed:
refactor: Remove unused methods

seizze committed:
refactor: Change to access data through closures

seizze committed:
refactor: Improve to use type beverage directly

seizze committed:
feat: Make fanta, georgia not to have enum

seizze committed:
docs: Update README.md

seizze commented:
> > 상속 관계에서 Equatable을 어떻게 적합하게 구현해야 할지 잘 떠오르지 않아서, 테스트 시에 같은지를 비교할 때 인스턴스가 같은지를 확인하는 === 연산자를 사용했는데, 이 방법이 맞는지 아니면 모든 타입들에 Equatable을 구현한 뒤에 테스트시에 타입 캐스팅등의 작업을 하여 == 연산자를 이용하는 것이 맞는지 잘 모르겠습니다.
> 
> 상속 관계를 가진 클래스들 Equatable 은 간단하게 생각하면 각 클래스 기준에서 모든 속성이 같으면 되겠죠.
> 인스턴스가 같은지 비교하는 것은 좋은 방법이 아닙니다. 그건 메모리 주소가 같은지 보는거니까 Equatable 아닙니다.
> 
> Hashable 프로토콜에 대해서도 학습해보세요. 모든 속성을 비교하는 것보다 효과적인 방법을 제시해줍니다.

Hashable 프로토콜을 이용하여 hashValue가 같은 경우 인스턴스가 같다고 판단하는 == 메서드를 추가하였습니다.

어려웠던 것은 서브클래스들의 경우에 추가된 프로퍼티의 값들도 같은지를 확인해야 하는데, 이 경우에 상위클래스의 비교 메서드가 호출되지 않고 서브클래스의 메서드가 호출되어 서브클래스에 있는 프로퍼티의 값들도 비교 해야 한다는 점이었습니다. `hash(into:)`를 오버라이드하여 하위클래스의 경우 오버라이드된 해시 함수가 호출되도록 하였습니다.

인스턴스가 같은지를 검사할 때, Beverage에 구현된 == 메서드가 호출되며, 각 인스턴스들마다 각자의 타입에 맞는 해시 함수가 호출되어 하위클래스들의 경우엔 추가된 프로퍼티까지 모두 검증할 수 있다는 것을 배웠습니다.

이를 이용하여 테스트 메서드들에서도 인스턴스를 비교하지 않고 == 연산자를 이용하도록 수정하였습니다.

그런데 궁금한 점은 다른 프로퍼티값들을 가지는 두 인스턴스의 경우에도 해시값이 같게 나오는 경우도 있다고 하는데 그 경우는 무시해도 되는 건가요?

seizze commented:
피드백 반영하였습니다!!

godrm commented:
> 그런데 궁금한 점은 다른 프로퍼티값들을 가지는 두 인스턴스의 경우에도 해시값이 같게 나오는 경우도 있다고 하는데 그 경우는 무시해도 되는 건가요?

말 그래도 두 인스턴스를 같은걸로 봐도 되느냐 아니냐가 중요한 포인트죠. 
그건 구현하는 모델이 어떤 대상을 모델링 했느냐에 따라 다릅니다. 상품을 구매하는 경우라면 같아야 할까요? 달라야 할까요? 그 기준에 따라 다릅니다. 더 정교하게 비교해야 하면 더 많은 값을, 혹은 더 큰 값을 비교해야겠죠

seizze commented:
음 아직 이해가 안갑니다 😭 그럼 프로퍼티 값은 다르지만 해시 충돌이 일어나는 경우도 같다고 봐도 되는것인가요?

godrm commented:
> 음 아직 이해가 안갑니다 😭 그럼 프로퍼티 값은 다르지만 해시 충돌이 일어나는 경우도 같다고 봐도 되는것인가요?

프로퍼티 값이 다른 데 해시 충돌이 일어나서 같다고 본다면 이상한 게 맞습니다. 적어도 그런 경우가 언제 발생하는지, 피해갈 수 있다면 피해가는 게 좋겠죠. 예를 들어 해시 버킷 크기가 65536개만 표현할 수 있다면 적어도 그정도 내에서는 구분이 되어야겠죠. 음료를 그 이상 넣을 수 있으냐 없느냐를 확인해봐야 하는거죠. 그정도 이상 들어가서 비교할 일이 없으면 안전한거구요. 아니면 다른 방법을 찾아야죠. 
모든 프로그램은 제약 사항이 있습니다. 그걸 기준으로 정하는 걸 엔지니어는 데이터 구조와 함께 판단해야 합니다.

seizze commented:
아하 넵 이해했습니다! 답변 감사합니다 😆
